Add channel plugin for state of channels and users

nitori opened the pull request:
Keeps track of currently joined channel and it's users.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -0,0 +1,305 @@
+        ctx.users[lnew_nick].nickname = new_nick
+        del ctx.users[lnick]
+
+    # should we do this here?
```
I wonder. It makes more sense to include it in the place where `Network.nickname` is being tracked. Maybe a separate "identity" core plugin, which would also take care of nick collisions on join (thus currently happens in `network.py` directly).

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -0,0 +1,305 @@
+        self.modes = modes
+
+
+class User(_Base):
```
Isn't this exactly what `shangahi.irc.Prefix` is for? They are identical, except that Prefix's `__init__` has no defaults, which can be added easily.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -0,0 +1,305 @@
+__plugin_description__ = 'Track channel state'
+
+
+class _Base:
```
What is this for? Aren't all these classes basically namedtuples anyway?

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -0,0 +1,305 @@
+        return hash(id(self))
+
+
+class Channel(_Base):
```
I wonder whether this should go into `shanghai.irc`.

FichteFoll commented:
Here is how I would imagine the channel API to work, with the example of reimplementing your two commands from `test.py`:

```py
    @ctx.channel_event(ChannelEventName.MESSAGE)
    def on_chanmsg(ctx: ChannelContext, message: Message):
        line = message.params[-1]
        words = line.split()
        
        if words[0] == '!nicks':
            ctx.say(' '.join(dehighlight(u.name) for u in ctx.users))

        elif words[0] == '!channels':
            ctx.say(', '.join(f"{c.name} ({len(c.users)})" for c in ctx.network.channels))
```

nitori commented on shanghai/core_plugins/channel.py:
```diff
@@ -0,0 +1,305 @@
+
+    ctx.add_method('remove_nick_from_channel', remove_nick_from_channel)
+
+    ctx.add_method('nick_lower', nick_lower)
```
That, I did not know. I shall change it.

nitori commented on shanghai/core_plugins/channel.py:
```diff
@@ -0,0 +1,305 @@
+
+@global_event(GlobalEventName.INIT_NETWORK_CTX)
+async def init_context(ctx: NetworkContext):
+    ctx.add_attribute('channels', {})  # l(channel-name) -> channel
```
Wasn't meant as type annotation, but rather a simple not, what value is expected for the key (in this case a lowercase version of the channel name)

I'm still kinda unsure about the dict thing. I thought it might as well work to just use `__hash__` and make it return the lowercase version of the channel name and then just use a `set`. Same for users and possibly joins.

nitori commented:
> Also, RFC 2811 uses the term "members" instead of "users" (which most irc clients use, I think).

The term "members" kinda means "members of a channel", no? But the users dict doesn't reflect that. The users dict contains "all visible members of all channels without duplicates". If anything you could probably convince me to rename "joins" to "memberships" or something like that, and later have a "get_members" API, or similar, on the channel context, once it's there.

Once again... I really dislike the function nesting.

```python
    @n_ctx.channel_event(ChannelEventName.MESSAGE)
    def on_chanmsg(c_ctx: ChannelContext, message: Message):
        …
```
This kinda implies it's called from within a function, right? Because `n_ctx` wouldn't be available. I understand decorators are great 'n stuff, but I *really* dislike it, when that kind of semi-required nesting is part of the API for registering event callbacks (which is why I "misused" the decorator in this plugin to register functions on top-level). And making the network context available from within `on_chanmsg` is just a matter of passing it to the channel context object.

nitori commented on shanghai/core_plugins/channel.py:
```diff
@@ -0,0 +1,305 @@
+    return ctx.nick_lower(chan)
+
+
+def nick_cmp(ctx: NetworkContext, nick1: str, nick2: str):
```
Oh right. I could have named them that way.

FichteFoll commented:
>The term "members" kinda means "members of a channel", no? But the users dict doesn't reflect that. 

I know it doesn't do that currently.

But realistically, you don't care about particular users, or if you do, you don't care about which channels they are in because you'd communicate with them directly via PRIVMSG. That means we have the semantic "Server <-1--n-> User" and "Server <-1--n-> Channel <-1--n-> User" relationships, even though technically Channel and User are many-to-many. The point is, however, that we only truly know which channel has which users and not which users is on which channels, because we only know of a limited subset of available channels, so a one-user-to-many-channels relationship would only be half true.

---

>This kinda implies it's called from within a function, right? Because n_ctx wouldn't be available.

Not sure I follow with the "function" part. 

`n_ctx` is carried over from the `init_network_context`, which is basically an "init this plugin" routine,  as a clojure. If we were using an inheritation-based approach instead of a functional one, you could think of this as the class's `__init__` with each variable definition equating to an attribute assignment to `self`.

In other frameworks you would probably do this initialization on a module-global level, but because we have multi-network functionality as a goal, with different plugins and plugin settings for each, we can't have plugins share network-specific state globally.
As such, a `NetworkContext` object will *always* be available in any event handlers, because it is required to register any events on it in the first place. The `ChannelContext` would just be a layer on top of the network context for events that are specific for a particular channel.

---

I hope I'm making sense here.

nitori commented:
This is what I meant. Taken from `test.py`

```python
@global_event(GlobalEventName.INIT_NETWORK_CTX)
async def init_context(ctx):
    @ctx.message_event('PRIVMSG')
    async def on_privmsg(ctx, message):
        line = message.params[-1]
        words = line.split()
        […]
```

And expanding it on your example, wouldn't it look something like this?

```python
@global_event(GlobalEventName.INIT_NETWORK_CTX)
async def init_context(n_ctx: NetworkContext):
    @n_ctx.channel_event(ChannelEventName.MESSAGE)
    def on_chanmsg(c_ctx: ChannelContext, message: Message):
        line = message.params[-1]
        words = line.split()
```

In the channel plugin I instead used something like this:

```python
def on_chanmsg(c_ctx: ChannelContext, message: Message):
    line = message.params[-1]
    words = line.split()
    # and if network context is required, I'd suggest just accessing it via
    # c_ctx, e.g. c_ctx.network_context or the like.

@global_event(GlobalEventName.INIT_NETWORK_CTX)
async def init_context(n_ctx: NetworkContext):
    n_ctx.channel_event(ChannelEventName.MESSAGE)(on_chanmsg)
    # "misused" the decorator to register on_chanmsg module level function instead of closure.
```

It would certainly be possible to use a decorator on module level functions as well. I assume `init_context` would only be called, if the plugin is enabled for said network. So using a module level decorator could store the module level functions in a (global) cache for all the functions that desire to be registered to certain events *as if* they where registered in `init_context`, except it's handled by the same routine that dispatches `init_context`, e.g. pulls out those "desiring" functions, registers them with the specific events (shortly before or after the `init_context` dispatch, and all with the same "don't execute for this network" logic - if necessary - behind it.

Well, that's not necessary though. Registering everything in `init_context` is fine, though I would prefer an additional non-decorator API for that as well.

FichteFoll commented:
> And expanding it on your example, wouldn't it look something like this?

Yes, I just skipped that part and indeted it because I though it was a given.

Your method of specifying the methods globally and just registering them in the init_context is fine and perfectly valid. It loses the advantage of encapsulating network-specific variables through clojures though, which is why you resorted to setting the attributes on `ctx` directly (which I dislike). In any case, "private" variables that are not supposed to be read by other plugins should be prefixed with an underscore.

I used it myself in the [`ctcp.py` plugin](https://github.com/chireiden/shanghai/blob/d1d6d25dfb53bff226aedd380e5db00be51ec14c/shanghai/core_plugins/ctcp.py#L58-L69). Though I probably would only use it for `ctx.add_method`, because imo it's better if the event handler and the event they are listening to are more closely together in code (a decorator would be directly above the code).

>  and if network context is required, I'd suggest just accessing it via c_ctx, e.g. c_ctx.network_context or the like.

I thought about this, but am undecided. We might as well just do it and see how things go.

> So using a module level decorator could store the module level functions in a (global) cache for all the functions that desire to be registered to certain events as if they where registered in init_context

Yes, that is possible as well. It probably makes for a "cleaner API" at first sight, but has the downside of being less flexible and being more troublesome to work with since more magic is involved in the background.
It would ease un- and reloading plugins, because we know which events they are listening to, but in order to allow dynamic event hooks we'd have to do that for the current architecture somehow anyway.

Your main critique is that event registration currently is too nested, correct?

nitori commented:
So, the latest changes basically implement new events for channel/private messages/notices and the channel context. In `test.py` plugin you can see how I used it.

I used non-caps event names for this (e.g. `ChannelMessage`), to indicate them being more abstract and not part of the protocol. Though I don't mind them changing.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -19,10 +19,11 @@
-from shanghai.event import global_event, GlobalEventName
-from shanghai.network import NetworkContext
+from shanghai.event import global_event, GlobalEventName, EventDispatcher, Priority
+from shanghai.network import ShadowAttributesMixin, NetworkContext
```
`ShadowAttributesMixin` is defined in `shanghai.util`.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -19,10 +19,11 @@
+from shanghai.event import global_event, GlobalEventName, EventDispatcher, Priority
+from shanghai.network import ShadowAttributesMixin, NetworkContext
 from shanghai.irc.server_reply import ServerReply
 from .message import Message
```
This import should probably be separated from the others. Otherwise, it should be made absolute as well.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -55,6 +56,78 @@ def __init__(self, channel: Channel, user: User, info: dict=None):
+        self.logger = logger
+
+    def say(self, message):
+        self.network_context.send_cmd('PRIVMSG', self._message.channel, message)
```
Just use `NetworkContext.send_msg` here, as added by `message.py`.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -55,6 +56,78 @@ def __init__(self, channel: Channel, user: User, info: dict=None):
+    def say(self, message):
+        self.network_context.send_cmd('PRIVMSG', self._message.channel, message)
+
+    def msg(self, target, message):
```
This already exists as `NetworkContext.send_msg` and is not channel-specific.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -55,6 +56,78 @@ def __init__(self, channel: Channel, user: User, info: dict=None):
+
+class ChannelMessage(BaseMessage):
+    def __repr__(self):
+        return f'<{self.__class__.__name__} type={self.command!r} ' \
```
PEP-8 recommends wrapping expressions in parentheses instead of using backslashes:

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -55,6 +56,78 @@ def __init__(self, channel: Channel, user: User, info: dict=None):
+class ChannelMessage(BaseMessage):
+    def __repr__(self):
+        return f'<{self.__class__.__name__} type={self.command!r} ' \
+               f'sender={self.prefix.name!r} source={self.params[0]!r} ' \
```
`source` is not an actual attribute of `ChannelMessage` (or `BaseMessage`). Neither is `message`.

A `__init__` should be implemented.

The same applies to `PrivateMessage`.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -286,7 +359,7 @@ def chan_eq(ctx: NetworkContext, chan1: str, chan2: str):
 
 
-@global_event(GlobalEventName.INIT_NETWORK_CTX)
+@global_event(GlobalEventName.INIT_NETWORK_CTX, priority=Priority.POST_CORE)
```
What does this init routine do that needs to happen after "core"?

nitori replied:
It needs to run before "DEFAULT" (so normal plugins can use this) and after "PRE_CORE" (message.py).

FichteFoll replied:
Just use `CORE` then.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -308,3 +381,36 @@ def chan_eq(ctx: NetworkContext, chan1: str, chan2: str):
+        return channel_contexts[lchannel]
+
+    @ctx.message_event('PRIVMSG')
+    @ctx.message_event('NOTICE')
```
By the way, using multiple decorators for the same function is intended to work but not well-tested at the moment. Specifically with regard to unregistering. Feel free to use it, however, since it's on my radar.

FichteFoll commented on shanghai/core_plugins/channel.py:
```diff
@@ -55,6 +56,78 @@ def __init__(self, channel: Channel, user: User, info: dict=None):
+    pass
+
+
+class PrivateNotice(PrivateMessage):
```
Wonder if we should call these just "Notice". It's the more often used term but "PrivateNotice" would follow the same naming as the *Message classes.

nitori replied:
I'm fine with using "Notice" here. Maybe even use "Query" instead of "PrivateMessage"? (or not..)

FichteFoll replied:
I always thought of "Query" as something you show the messages with someone in, i.e. a window. I can see how one could see it as a message in itself, though. Imo "PrivateMessage" is more appropriate.

FichteFoll commented:
Things are getting a bit messy now. I'm not so sure which comments of the first review still need addressing, but it wouldn't be hard to spot them again.

Either way, this looks much better now. 👍 

Would you mind commenting on my other question earlier?

> Your main critique is that event registration currently is too nested, correct?

nitori commented:
> Your main critique is that event registration currently is too nested, correct?

Well I wouldn't say it's "too nested", but yes. In general I don't like the way closures are registered (though I did end up using it myself in `channel.py`).

nitori commented:
Forgot to address the following:

> But realistically, you don't care about particular users, or if you do, you don't care about which channels they are in because you'd communicate with them directly via PRIVMSG. That means we have the semantic "Server <-1--n-> User" and "Server <-1--n-> Channel <-1--n-> User" relationships, even though technically Channel and User are many-to-many. The point is, however, that we only truly know which channel has which users and not which users is on which channels, because we only know of a limited subset of available channels, so a one-user-to-many-channels relationship would only be half true.

Well, commands like "!notify" or "!seen" might be interested in knowing in what channels are user is. At least from the bots perspective. I know that the bot will never have the full many-to-many information, because it will never on all channels and even then possibly never see all users (e.g. NickServ is usually hidden).

But I still think this is a better approach than having a set of users/members in each channel and having to "go through all channels" to find a specific user for e.g. the mentioned commands above. Or at least, I don't see any advantage/disadvantage of one of them over the other.

FichteFoll commented:
>Well, commands like "!notify" or "!seen" might be interested in knowing in what channels are user is. 

!notify, not really. You either receive a join event you can listen to or you /whois the user to see if they are online at the time of the command's invocation. You could take a shortcut by checking the channels you know of, still.

!seen, I wonder. First, it would depend on what the command is supposed to do exactly. Should it consider only the current channel? Then it doesn't matter. Should it consider all channels? Then, if the person is online, you'd see that in /whois. If they are not and had left or quit previously, you would have recorded that in a database.

That said, I'm okay with storing this data in a pseudo-relational database, although it seems like over-engineering because we'd never need it – and if we ever did, we could still iterate over all channels (just more slowly). Heck, we could even make it an actual database using in-memory sqlite.

FichteFoll commented:
>In general I don't like it that closures are registered instead of module level functions

I'll have to make a short write-up for this and ask a few other people about their opinion.

FichteFoll commented:
I address all my remaining change requests (and more) in #47.

(also note to self)
